var/spack/repos/builtin/packages/py-smartredis/package.py

```diff
@@ -0,0 +1,56 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PySmartredis(PythonPackage):
+    """A Python Interface for the SmartRedis Library Client"""
+
+    homepage = "https://www.craylabs.org/docs/smartredis.html"
+    pypi = "smartredis/smartredis-0.4.0.tar.gz"
+    git = "https://github.com/CrayLabs/SmartRedis"
+
+    maintainers("MattToast")
+
+    version("0.4.1", sha256="fff16ed1eb09648ac3c3f845373beb37f3ffe7414d8745ae36af9daf585f8c5b")
+    version("0.4.0", sha256="d12779aa8bb038e837c25eac41b178aab9e16b729d50ee360b5af8f813d9f1dd")
+
+    depends_on("python@3.7:3.10", type=("build", "run"))
+    depends_on("py-setuptools@42:", type=("build",))
+
+    depends_on("cmake@3.13:", type=("build",))
+
+    # Documented dependencies. Allow them to float according to semeantic versioning
+    # rules for easier concretization
+    depends_on("hiredis@1.1.0:1", type=("build", "link", "run"), when="@0.4.1")
+    depends_on("hiredis@1.0.0:1", type=("build", "link", "run"), when="@0.4.0")
+    depends_on("redis-plus-plus@1.3:1 cxxstd=17", type=("build", "link"))
+
+    # Unlisted dependency needed to build the python client. The pybind requirement
+    # can be found:
+    #  - in the `build-scripts/build_deps.sh` for SmartRedis <= v0.4.0
+    #  - in the `Makefile` under the `pybind` target for SmartRedis >= v0.4.1
+    # This dependency is intentionally left looser than the one used by SmartRedis
+    # as it does not use the features of the newer versions of pybind and allows for
+    # easier concretization
+    depends_on("py-pybind11@2.6:2", type=("build",))
```

Comment:
Where do the lower and upper bound come from?

Reply:
The lower bound comes from my own experimentation. For full disclosure, I'm trying to add this package so that I may create a package for [SmartSim](https://github.com/CrayLabs/SmartSim), a companion library that uses SmartRedis as a dependency. SmartSim's current "preferred" version of PyTorch is v1.11.0 (although it also can support newer versions) which relies on PyBind11 v2.6.

During my experimentation I was running into un-resolvable concretization problem unless I relaxed this constraint to v2.6 for the lower bound. While SmartRedis officially uses v2.10 in their own build system, it does not rely on any new features or bug fixes not already present in v2.6, so I felt it was okay to relax the constraint. Feel free to let me know if this feels like too much of a caveat to allow in and I would be willing to pin it the later version.

The upper bound comes from my misunderstanding of the last piece of feedback about relaxing non-pip-checked dependency version constraints. I am more that willing to remove that!!

Reply:
Personally I don't add any version constraints at all unless I know for a fact that older/newer versions won't work.

Reply:
Oh! Well in that case, SR should be able to compile with any relatively modern version of pybind11 (certainly with any of the versions currently offered through spack). I will remove both the lower and upper constraints!